yanix v0.12.0 (/tmp/wasmtime/crates/wasi-common/yanix) fails to compile on armv7-unknown-linux-gnueabihf

git clone the repo, cd wasmtime/crates/wasi-common/yanix and execute cargo build fails with this beauty: 

```
crates/wasi-common/yanix $ cargo build
    Updating crates.io index
  Downloaded cfg-if v0.1.10
  Downloaded thiserror v1.0.11
  Downloaded libc v0.2.67
  Downloaded log v0.4.8
  Downloaded bitflags v1.2.1
  Downloaded thiserror-impl v1.0.11
  Downloaded proc-macro2 v1.0.8
  Downloaded quote v1.0.2
  Downloaded unicode-xid v0.2.0
  Downloaded syn v1.0.15
   Compiling proc-macro2 v1.0.8
   Compiling unicode-xid v0.2.0
   Compiling syn v1.0.15
   Compiling bitflags v1.2.1
   Compiling libc v0.2.67
   Compiling log v0.4.8
   Compiling cfg-if v0.1.10
   Compiling quote v1.0.2
   Compiling thiserror-impl v1.0.11
   Compiling thiserror v1.0.11
   Compiling yanix v0.12.0 (/home/tim/wasmtime/crates/wasi-common/yanix)
error[E0308]: try expression alternatives have incompatible types
   --> crates/wasi-common/yanix/src/file.rs:247:23
    |
247 |     let offset: i64 = Errno::from_result(libc::lseek(fd, 0, libc::SEEK_CUR))?;
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                       |
    |                       expected `i64`, found `i32`
    |                       help: you can convert an `i32` to `i64`: `Errno::from_result(libc::lseek(fd, 0, libc::SEEK_CUR))?.into()`

error[E0277]: the trait bound `i32: std::convert::From<i64>` is not satisfied
  --> crates/wasi-common/yanix/src/sys/linux/mod.rs:10:17
   |
10 |         Ok(Self(loc))
   |                 ^^^ the trait `std::convert::From<i64>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<std::num::NonZeroI32>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<i32>` for `i64`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `yanix`.

```

aarch64 seems fine. 

rustc is v1.41.1 with llvm-9

same compile with verbose switch: 

```
cargo build --verbose
       Fresh unicode-xid v0.2.0
       Fresh cfg-if v0.1.10
       Fresh proc-macro2 v1.0.8
       Fresh quote v1.0.2
       Fresh libc v0.2.67
       Fresh log v0.4.8
       Fresh bitflags v1.2.1
       Fresh syn v1.0.15
       Fresh thiserror-impl v1.0.11
       Fresh thiserror v1.0.11
   Compiling yanix v0.12.0 (/home/tim/wasmtime/crates/wasi-common/yanix)
     Running `rustc --edition=2018 --crate-name yanix crates/wasi-common/yanix/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=0f2083b39d9868a9 -C extra-filename=-0f2083b39d9868a9 --out-dir /home/tim/wasmtime/target/debug/deps -C incremental=/home/tim/wasmtime/target/debug/incremental -L dependency=/home/tim/wasmtime/target/debug/deps --extern bitflags=/home/tim/wasmtime/target/debug/deps/libbitflags-2f0d650362f6563c.rmeta --extern cfg_if=/home/tim/wasmtime/target/debug/deps/libcfg_if-73270e7e135e30d3.rmeta --extern libc=/home/tim/wasmtime/target/debug/deps/liblibc-e4e1be7f8b2f2e45.rmeta --extern log=/home/tim/wasmtime/target/debug/deps/liblog-e99f890746485ea4.rmeta --extern thiserror=/home/tim/wasmtime/target/debug/deps/libthiserror-2371134c478ad8d8.rmeta`
error[E0308]: try expression alternatives have incompatible types
   --> crates/wasi-common/yanix/src/file.rs:247:23
    |
247 |     let offset: i64 = Errno::from_result(libc::lseek(fd, 0, libc::SEEK_CUR))?;
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                       |
    |                       expected `i64`, found `i32`
    |                       help: you can convert an `i32` to `i64`: `Errno::from_result(libc::lseek(fd, 0, libc::SEEK_CUR))?.into()`

error[E0277]: the trait bound `i32: std::convert::From<i64>` is not satisfied
  --> crates/wasi-common/yanix/src/sys/linux/mod.rs:10:17
   |
10 |         Ok(Self(loc))
   |                 ^^^ the trait `std::convert::From<i64>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<std::num::NonZeroI32>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<i32>` for `i64`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `yanix`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name yanix crates/wasi-common/yanix/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=0f2083b39d9868a9 -C extra-filename=-0f2083b39d9868a9 --out-dir /home/tim/wasmtime/target/debug/deps -C incremental=/home/tim/wasmtime/target/debug/incremental -L dependency=/home/tim/wasmtime/target/debug/deps --extern bitflags=/home/tim/wasmtime/target/debug/deps/libbitflags-2f0d650362f6563c.rmeta --extern cfg_if=/home/tim/wasmtime/target/debug/deps/libcfg_if-73270e7e135e30d3.rmeta --extern libc=/home/tim/wasmtime/target/debug/deps/liblibc-e4e1be7f8b2f2e45.rmeta --extern log=/home/tim/wasmtime/target/debug/deps/liblog-e99f890746485ea4.rmeta --extern thiserror=/home/tim/wasmtime/target/debug/deps/libthiserror-2371134c478ad8d8.rmeta` (exit code: 1)
```

ok, so I git cloned your fork and did `git checkout yanix-error-cleanup`, followed by `cd crates/wasi-common/yanix` and cargo build. It turns out that with the commits you added plus my patch from above, there's a new error now: 

```
cargo build --target armv7-unknown-linux-gnueabihf
   Compiling yanix v0.12.0 (/tmp/wasmtime/crates/wasi-common/yanix)
error[E0433]: failed to resolve: use of undeclared type or module `Errno`
   --> crates/wasi-common/yanix/src/file.rs:247:23
    |
247 |     let offset: i64 = Errno::from_result(libc::lseek(fd, 0, libc::SEEK_CUR))?.into();
    |                       ^^^^^ use of undeclared type or module `Errno`

error[E0277]: the trait bound `i32: std::convert::From<i64>` is not satisfied
  --> crates/wasi-common/yanix/src/sys/linux/mod.rs:10:17
   |
10 |         Ok(Self(loc))
   |                 ^^^ the trait `std::convert::From<i64>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<std::num::NonZeroI32>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<i32>` for `i64`

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0277, E0433.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `yanix`.
```

and I think that's due to you changing from Errno to Error in that line, right?

with the rebased source tree, and when #1231 is applied, the second error remains: 

```
error[E0277]: the trait bound `i32: std::convert::From<i64>` is not satisfied
  --> crates/wasi-common/yanix/src/sys/linux/mod.rs:10:17
   |
10 |         Ok(Self(loc))
   |                 ^^^ the trait `std::convert::From<i64>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<std::num::NonZeroI32>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<i32>` for `i64`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
```

I tried to switch to `from_raw(loc: i32)` in that line, since i64 seems to be wrong, but that doesn't seem to solve anything :/

still struggeling, five new regressions since https://github.com/bytecodealliance/wasmtime/commit/061390ee1b1085a2191019d2219da730ec62485f

```
/tmp/wasmtime/crates/wasi-common/yanix $ cargo build --target armv7-unknown-linux-gnueabihf
   Compiling libc v0.2.67
   Compiling cfg-if v0.1.10
   Compiling bitflags v1.2.1
   Compiling log v0.4.8
   Compiling filetime v0.2.8
   Compiling yanix v0.12.0 (/tmp/wasmtime/crates/wasi-common/yanix)
error[E0308]: mismatched types
  --> crates/wasi-common/yanix/src/filetime.rs:18:29
   |
18 |                     tv_sec: ft.seconds(),
   |                             ^^^^^^^^^^^^ expected `i32`, found `i64`

error[E0308]: mismatched types
  --> crates/wasi-common/yanix/src/filetime.rs:19:30
   |
19 |                     tv_nsec: i64::from(ft.nanoseconds()),
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found `i64`

error[E0308]: mismatched types
  --> crates/wasi-common/yanix/src/filetime.rs:72:22
   |
72 |             tv_nsec: UTIME_NOW,
   |                      ^^^^^^^^^ expected `i32`, found `i64`

error[E0308]: mismatched types
  --> crates/wasi-common/yanix/src/filetime.rs:76:22
   |
76 |             tv_nsec: UTIME_OMIT,
   |                      ^^^^^^^^^^ expected `i32`, found `i64`

error[E0277]: the trait bound `i32: std::convert::From<i64>` is not satisfied
  --> crates/wasi-common/yanix/src/sys/linux/mod.rs:13:17
   |
13 |         Ok(Self(loc))
   |                 ^^^ the trait `std::convert::From<i64>` is not implemented for `i32`
   |
   = help: the following implementations were found:
             <i32 as std::convert::From<bool>>
             <i32 as std::convert::From<i16>>
             <i32 as std::convert::From<i8>>
             <i32 as std::convert::From<std::num::NonZeroI32>>
           and 2 others
   = note: required because of the requirements on the impl of `std::convert::Into<i32>` for `i64`

error[E0308]: mismatched types
  --> crates/wasi-common/yanix/src/sys/linux/utimesat.rs:39:17
   |
39 |         tv_sec: ft.seconds(),
   |                 ^^^^^^^^^^^^ expected `i32`, found `i64`

error: aborting due to 6 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `yanix`.

To learn more, run the command again with --verbose.
```

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -120,18 +120,30 @@ jobs:
         RUST_BACKTRACE: 1
 
   # Install wasm32-unknown-emscripten target, and ensure `crates/wasi-common`
-  # compiles to Emscripten.
-  # TODO enable once rust-lang/rust#66308 is fixed
-  # emscripten:
-  #   name: Emscripten
-  #   runs-on: ubuntu-latest
-  #   steps:
-  #   - uses: actions/checkout@v1
-  #     with:
-  #       submodules: true
-  #   - uses: ./.github/actions/install-rust
-  #   - run: rustup target add wasm32-unknown-emscripten
-  #   - run: cargo build --target wasm32-unknown-emscripten -p wasi-common
+  # compiles to it.
+  emscripten:
+    name: Emscripten
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+      with:
+        submodules: true
+    - uses: ./.github/actions/install-rust
+    - run: rustup target add wasm32-unknown-emscripten
+    - run: cargo check --target wasm32-unknown-emscripten -p wasi-common
+  
+  # Install armv7-unknown-linux-gnueabihf target, and ensure `crates/wasi-common`
+  # compiles to it.
+  armv7:
+    name: armv7
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+      with:
+        submodules: true
+    - uses: ./.github/actions/install-rust
+    - run: rustup target add armv7-unknown-linux-gnueabihf
+    - run: cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
 
   # Perform all tests (debug mode) for `wasmtime`. This runs stable/beta/nightly
   # channels of Rust as well as macOS/Linux/Windows.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -119,31 +119,22 @@ jobs:
       env:
         RUST_BACKTRACE: 1
 
-  # Install wasm32-unknown-emscripten target, and ensure `crates/wasi-common`
-  # compiles to it.
-  emscripten:
-    name: Emscripten
+  # Check whether `crates/wasi-common` cross-compiles to the following targets:
+  # * wasm32-unknown-emscripten
+  # * armv7-unknown-linux-gnueabihf
+  crosscompile:
+    name: Cross-platform checks
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v1
       with:
         submodules: true
     - uses: ./.github/actions/install-rust
-    - run: rustup target add wasm32-unknown-emscripten
-    - run: cargo check --target wasm32-unknown-emscripten -p wasi-common
-  
-  # Install armv7-unknown-linux-gnueabihf target, and ensure `crates/wasi-common`
-  # compiles to it.
-  armv7:
-    name: armv7
-    runs-on: ubuntu-latest
-    steps:
-    - uses: actions/checkout@v1
-      with:
-        submodules: true
-    - uses: ./.github/actions/install-rust
-    - run: rustup target add armv7-unknown-linux-gnueabihf
-    - run: cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
+    - run: |
+        rustup target add wasm32-unknown-emscripten
+        cargo check --target wasm32-unknown-emscripten -p wasi-common
+        rustup target add armv7-unknown-linux-gnueabihf
+        cargo check --target armv7-unknown-linux-gnueabihf -p wasi-common
 
   # Perform all tests (debug mode) for `wasmtime`. This runs stable/beta/nightly
   # channels of Rust as well as macOS/Linux/Windows.

--- a/crates/wasi-common/src/old/snapshot_0/sys/unix/emscripten/host_impl.rs
+++ b/crates/wasi-common/src/old/snapshot_0/sys/unix/emscripten/host_impl.rs
@@ -1,15 +1,15 @@
-use crate::old::snapshot_0::{wasi, Result};
+use crate::old::snapshot_0::wasi::{self, WasiResult};
 
 pub(crate) const O_RSYNC: yanix::file::OFlag = yanix::file::OFlag::RSYNC;
 
-pub(crate) fn stdev_from_nix(dev: libc::dev_t) -> Result<wasi::__wasi_device_t> {
+pub(crate) fn stdev_from_nix(dev: libc::dev_t) -> WasiResult<wasi::__wasi_device_t> {
     Ok(wasi::__wasi_device_t::from(dev))
 }
 
-pub(crate) fn stino_from_nix(ino: libc::ino_t) -> Result<wasi::__wasi_inode_t> {
+pub(crate) fn stino_from_nix(ino: libc::ino_t) -> WasiResult<wasi::__wasi_inode_t> {
     Ok(wasi::__wasi_device_t::from(ino))
 }
 
-pub(crate) fn stnlink_from_nix(nlink: libc::nlink_t) -> Result<wasi::__wasi_linkcount_t> {
+pub(crate) fn stnlink_from_nix(nlink: libc::nlink_t) -> WasiResult<wasi::__wasi_linkcount_t> {
     Ok(nlink)
 }

--- a/crates/wasi-common/yanix/src/dir.rs
+++ b/crates/wasi-common/yanix/src/dir.rs
@@ -2,7 +2,8 @@ use crate::{
     file::FileType,
     sys::dir::{iter_impl, EntryImpl},
 };
-use std::io::Result;
+use std::convert::TryInto;
+use std::io::{Error, Result};
 use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
 use std::{ffi::CStr, io, ops::Deref, ptr};
 
@@ -97,6 +98,18 @@ pub struct SeekLoc(pub(crate) libc::c_long);
 impl SeekLoc {
     pub fn to_raw(&self) -> i64 {
         self.0.into()
+    }
+
+    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
+        // The cookie (or `loc`) is an opaque value, and applications aren't supposed to do
+        // arithmetic on them or pick their own values or have any awareness of the numeric
+        // range of the values. They're just supposed to pass back in the values that we
+        // give them. And any value we give them will be convertable back to `long`,
+        // because that's the type the OS gives them to us in. So return an `EINVAL`.
+        let loc = loc
+            .try_into()
+            .map_err(|_| Error::from_raw_os_error(libc::EINVAL))?;
+        Ok(Self(loc))
     }
 }
 

--- a/crates/wasi-common/yanix/src/filetime.rs
+++ b/crates/wasi-common/yanix/src/filetime.rs
@@ -10,16 +10,16 @@ use std::io::Result;
 
 pub use super::sys::filetime::*;
 
+/// Internal trait which specialises `filetime::FileTime`'s
+/// `seconds` and `nanoseconds` accessors for different
+/// pointer widths (32 and 64bit currently).
 pub(crate) trait FileTimeExt {
-    fn seconds_checked(&self) -> Result<libc::time_t>;
-    #[cfg(target_pointer_width = "32")]
-    fn nanoseconds_checked(&self) -> Result<i32>;
-    #[cfg(target_pointer_width = "64")]
-    fn nanoseconds_checked(&self) -> Result<i64>;
+    fn seconds_(&self) -> Result<libc::time_t>;
+    fn nanoseconds_(&self) -> libc::c_long;
 }
 
 impl FileTimeExt for filetime::FileTime {
-    fn seconds_checked(&self) -> Result<libc::time_t> {
+    fn seconds_(&self) -> Result<libc::time_t> {
         use std::convert::TryInto;
         use std::io::Error;
         let sec = match self.seconds().try_into() {
@@ -31,18 +31,13 @@ impl FileTimeExt for filetime::FileTime {
         };
         Ok(sec)
     }
-    #[cfg(target_pointer_width = "32")]
-    fn nanoseconds_checked(&self) -> Result<i32> {
+    fn nanoseconds_(&self) -> libc::c_long {
         use std::convert::TryInto;
         // According to [filetime] docs, since the nanoseconds value is always less than 1 billion,
-        // any value should be convertible to `i32`, hence we can `unwrap` outright.
+        // any value should be convertible to `libc::c_long`, hence we can `unwrap` outright.
         //
         // [filetime]: https://docs.rs/filetime/0.2.8/filetime/struct.FileTime.html#method.nanoseconds
-        Ok(self.nanoseconds().try_into().unwrap())
-    }
-    #[cfg(target_pointer_width = "64")]
-    fn nanoseconds_checked(&self) -> Result<i64> {
-        Ok(i64::from(self.nanoseconds()))
+        self.nanoseconds().try_into().unwrap()
     }
 }
 
@@ -75,8 +70,8 @@ pub(crate) fn to_timespec(ft: &FileTime) -> Result<libc::timespec> {
             tv_nsec: libc::UTIME_OMIT,
         },
         FileTime::FileTime(ft) => libc::timespec {
-            tv_sec: ft.seconds_checked()?,
-            tv_nsec: ft.nanoseconds_checked()?,
+            tv_sec: ft.seconds_()?,
+            tv_nsec: ft.nanoseconds_(),
         },
     };
     Ok(ts)

--- a/crates/wasi-common/yanix/src/filetime.rs
+++ b/crates/wasi-common/yanix/src/filetime.rs
@@ -11,23 +11,17 @@ use std::io::Result;
 pub use super::sys::filetime::*;
 
 pub(crate) trait FileTimeExt {
-    #[cfg(target_pointer_width = "32")]
-    fn seconds_checked(&self) -> Result<i32>;
+    fn seconds_checked(&self) -> Result<libc::time_t>;
     #[cfg(target_pointer_width = "32")]
     fn nanoseconds_checked(&self) -> Result<i32>;
-    #[cfg(target_pointer_width = "64")]
-    fn seconds_checked(&self) -> Result<i64>;
     #[cfg(target_pointer_width = "64")]
     fn nanoseconds_checked(&self) -> Result<i64>;
 }
 
 impl FileTimeExt for filetime::FileTime {
-    #[cfg(target_pointer_width = "32")]
-    fn seconds_checked(&self) -> Result<i32> {
+    fn seconds_checked(&self) -> Result<libc::time_t> {
         use std::convert::TryInto;
         use std::io::Error;
-        // 32bit OS expects `tv_sec` field to be `i32`.
-        // Perform a checked conversion, log error if any, and convert to libc::EOVERFLOW.
         let sec = match self.seconds().try_into() {
             Ok(sec) => sec,
             Err(_) => {
@@ -40,21 +34,11 @@ impl FileTimeExt for filetime::FileTime {
     #[cfg(target_pointer_width = "32")]
     fn nanoseconds_checked(&self) -> Result<i32> {
         use std::convert::TryInto;
-        use std::io::Error;
-        // 32bit OS expects `tv_nsec` field to be `i32`.
-        // Perform a checked conversion, log error if any, and convert to libc::EOVERFLOW.
-        let nsec = match self.nanoseconds().try_into() {
-            Ok(nsec) => nsec,
-            Err(_) => {
-                log::debug!("filetime_to_timespec failed converting nanoseconds to required width");
-                return Err(Error::from_raw_os_error(libc::EOVERFLOW));
-            }
-        };
-        Ok(nsec)
-    }
-    #[cfg(target_pointer_width = "64")]
-    fn seconds_checked(&self) -> Result<i64> {
-        Ok(self.seconds())
+        // According to [filetime] docs, since the nanoseconds value is always less than 1 billion,
+        // any value should be convertible to `i32`, hence we can `unwrap` outright.
+        //
+        // [filetime]: https://docs.rs/filetime/0.2.8/filetime/struct.FileTime.html#method.nanoseconds
+        Ok(self.nanoseconds().try_into().unwrap())
     }
     #[cfg(target_pointer_width = "64")]
     fn nanoseconds_checked(&self) -> Result<i64> {
@@ -84,11 +68,11 @@ pub(crate) fn to_timespec(ft: &FileTime) -> Result<libc::timespec> {
     let ts = match ft {
         FileTime::Now => libc::timespec {
             tv_sec: 0,
-            tv_nsec: UTIME_NOW,
+            tv_nsec: libc::UTIME_NOW,
         },
         FileTime::Omit => libc::timespec {
             tv_sec: 0,
-            tv_nsec: UTIME_OMIT,
+            tv_nsec: libc::UTIME_OMIT,
         },
         FileTime::FileTime(ft) => libc::timespec {
             tv_sec: ft.seconds_checked()?,

--- a/crates/wasi-common/yanix/src/filetime.rs
+++ b/crates/wasi-common/yanix/src/filetime.rs
@@ -10,40 +10,55 @@ use std::io::Result;
 
 pub use super::sys::filetime::*;
 
-cfg_if::cfg_if! {
-    if #[cfg(not(target_os = "emscripten"))] {
-        fn filetime_to_timespec(ft: &filetime::FileTime) -> Result<libc::timespec> {
-            Ok(
-                libc::timespec {
-                    tv_sec: ft.seconds(),
-                    tv_nsec: i64::from(ft.nanoseconds()),
-                }
-            )
-        }
-    } else {
-        fn filetime_to_timespec(ft: &filetime::FileTime) -> Result<libc::timespec> {
-            use std::convert::TryInto;
-            use std::io::Error;
-            // Emscripten expects both `tv_sec` and `tv_nsec` fields to be `i32`.
-            // Here however `ft.seconds() -> i64` and `ft.nanoseconds() -> u32` so
-            // a simple `as` cast may be insufficient. So, perform a checked conversion,
-            // log error if any, and convert to libc::EOVERFLOW.
-            let tv_sec = match ft.seconds().try_into() {
-                Ok(sec) => sec,
-                Err(_) => {
-                    log::debug!("filetime_to_timespec failed converting seconds to required width");
-                    return Err(Error::from_raw_os_error(libc::EOVERFLOW));
-                }
-            };
-            let tv_nsec = match ft.nanoseconds().try_into() {
-                Ok(nsec) => nsec,
-                Err(_) => {
-                    log::debug!("filetime_to_timespec failed converting nanoseconds to required width");
-                    return Err(Error::from_raw_os_error(libc::EOVERFLOW));
-                }
-            };
-            Ok(libc::timespec { tv_sec, tv_nsec })
-        }
+pub(crate) trait FileTimeExt {
+    #[cfg(target_pointer_width = "32")]
+    fn seconds_checked(&self) -> Result<i32>;
+    #[cfg(target_pointer_width = "32")]
+    fn nanoseconds_checked(&self) -> Result<i32>;
+    #[cfg(target_pointer_width = "64")]
+    fn seconds_checked(&self) -> Result<i64>;
+    #[cfg(target_pointer_width = "64")]
+    fn nanoseconds_checked(&self) -> Result<i64>;
+}
+
+impl FileTimeExt for filetime::FileTime {
+    #[cfg(target_pointer_width = "32")]
+    fn seconds_checked(&self) -> Result<i32> {
+        use std::convert::TryInto;
+        use std::io::Error;
+        // 32bit OS expects `tv_sec` field to be `i32`.
+        // Perform a checked conversion, log error if any, and convert to libc::EOVERFLOW.
+        let sec = match self.seconds().try_into() {
+            Ok(sec) => sec,
+            Err(_) => {
+                log::debug!("filetime_to_timespec failed converting seconds to required width");
+                return Err(Error::from_raw_os_error(libc::EOVERFLOW));
+            }
+        };
+        Ok(sec)
+    }
+    #[cfg(target_pointer_width = "32")]
+    fn nanoseconds_checked(&self) -> Result<i32> {
+        use std::convert::TryInto;
+        use std::io::Error;
+        // 32bit OS expects `tv_nsec` field to be `i32`.
+        // Perform a checked conversion, log error if any, and convert to libc::EOVERFLOW.
+        let nsec = match self.nanoseconds().try_into() {
+            Ok(nsec) => nsec,
+            Err(_) => {
+                log::debug!("filetime_to_timespec failed converting nanoseconds to required width");
+                return Err(Error::from_raw_os_error(libc::EOVERFLOW));
+            }
+        };
+        Ok(nsec)
+    }
+    #[cfg(target_pointer_width = "64")]
+    fn seconds_checked(&self) -> Result<i64> {
+        Ok(self.seconds())
+    }
+    #[cfg(target_pointer_width = "64")]
+    fn nanoseconds_checked(&self) -> Result<i64> {
+        Ok(i64::from(self.nanoseconds()))
     }
 }
 
@@ -75,7 +90,10 @@ pub(crate) fn to_timespec(ft: &FileTime) -> Result<libc::timespec> {
             tv_sec: 0,
             tv_nsec: UTIME_OMIT,
         },
-        FileTime::FileTime(ft) => filetime_to_timespec(ft)?,
+        FileTime::FileTime(ft) => libc::timespec {
+            tv_sec: ft.seconds_checked()?,
+            tv_nsec: ft.nanoseconds_checked()?,
+        },
     };
     Ok(ts)
 }

--- a/crates/wasi-common/yanix/src/sys/bsd/dir.rs
+++ b/crates/wasi-common/yanix/src/sys/bsd/dir.rs
@@ -57,10 +57,3 @@ impl EntryExt for Entry {
         Ok(self.0.loc)
     }
 }
-
-impl SeekLoc {
-    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
-        let loc = loc.into();
-        Ok(Self(loc))
-    }
-}

--- a/crates/wasi-common/yanix/src/sys/bsd/filetime.rs
+++ b/crates/wasi-common/yanix/src/sys/bsd/filetime.rs
@@ -2,33 +2,10 @@
 //! with setting the file times specific to BSD-style *nixes.
 use crate::filetime::FileTime;
 use crate::from_success_code;
-use cfg_if::cfg_if;
 use std::ffi::CStr;
 use std::fs::File;
 use std::io::Result;
 use std::sync::atomic::{AtomicUsize, Ordering::SeqCst};
-
-cfg_if! {
-    if #[cfg(any(
-            target_os = "macos",
-            target_os = "freebsd",
-            target_os = "ios",
-            target_os = "dragonfly"
-    ))] {
-        pub(crate) const UTIME_NOW: i64 = -1;
-        pub(crate) const UTIME_OMIT: i64 = -2;
-    } else if #[cfg(target_os = "openbsd")] {
-        // These are swapped compared to macos, freebsd, ios, and dragonfly.
-        // https://github.com/openbsd/src/blob/master/sys/sys/stat.h#L187
-        pub(crate) const UTIME_NOW: i64 = -2;
-        pub(crate) const UTIME_OMIT: i64 = -1;
-    } else if #[cfg(target_os = "netbsd" )] {
-        // These are the same as for Linux.
-        // http://cvsweb.netbsd.org/bsdweb.cgi/src/sys/sys/stat.h?rev=1.69&content-type=text/x-cvsweb-markup&only_with_tag=MAIN
-        pub(crate) const UTIME_NOW: i64 = 1_073_741_823;
-        pub(crate) const UTIME_OMIT: i64 = 1_073_741_822;
-    }
-}
 
 /// Wrapper for `utimensat` syscall, however, with an added twist such that `utimensat` symbol
 /// is firstly resolved (i.e., we check whether it exists on the host), and only used if that is

--- a/crates/wasi-common/yanix/src/sys/emscripten/filetime.rs
+++ b/crates/wasi-common/yanix/src/sys/emscripten/filetime.rs
@@ -5,9 +5,6 @@ use crate::from_success_code;
 use std::fs::File;
 use std::io::Result;
 
-pub(crate) const UTIME_NOW: i32 = 1_073_741_823;
-pub(crate) const UTIME_OMIT: i32 = 1_073_741_822;
-
 /// Wrapper for `utimensat` syscall. In Emscripten, there is no point in dynamically resolving
 /// if `utimensat` is available as it always was and will be.
 pub fn utimensat(

--- a/crates/wasi-common/yanix/src/sys/emscripten/mod.rs
+++ b/crates/wasi-common/yanix/src/sys/emscripten/mod.rs
@@ -5,21 +5,3 @@ pub(crate) mod fadvise;
 #[path = "../linux/file.rs"]
 pub(crate) mod file;
 pub(crate) mod filetime;
-
-use crate::dir::SeekLoc;
-use std::convert::TryInto;
-use std::io::{Error, Result};
-
-impl SeekLoc {
-    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
-        // The cookie (or `loc`) is an opaque value, and applications aren't supposed to do
-        // arithmetic on them or pick their own values or have any awareness of the numeric
-        // range of the values. They're just supposed to pass back in the values that we
-        // give them. And any value we give them will be convertable back to `long`,
-        // because that's the type the OS gives them to us in. So return an `EINVAL`.
-        let loc = loc
-            .try_into()
-            .map_err(|_| Error::from_raw_os_error(libc::EINVAL))?;
-        Ok(Self(loc))
-    }
-}

--- a/crates/wasi-common/yanix/src/sys/linux/filetime.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/filetime.rs
@@ -6,16 +6,6 @@ use std::fs::File;
 use std::io::Result;
 use std::sync::atomic::{AtomicBool, Ordering::Relaxed};
 
-cfg_if::cfg_if! {
-    if #[cfg(target_pointer_width = "64")] {
-        pub(crate) const UTIME_NOW: i64 = 1_073_741_823;
-        pub(crate) const UTIME_OMIT: i64 = 1_073_741_822;
-    } else {
-        pub(crate) const UTIME_NOW: i32 = 1_073_741_823;
-        pub(crate) const UTIME_OMIT: i32 = 1_073_741_822;
-    }
-}
-
 /// Wrapper for `utimensat` syscall, however, with an added twist such that `utimensat` symbol
 /// is firstly resolved (i.e., we check whether it exists on the host), and only used if that is
 /// the case. Otherwise, the syscall resorts to a less accurate `utimesat` emulated syscall.

--- a/crates/wasi-common/yanix/src/sys/linux/filetime.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/filetime.rs
@@ -6,8 +6,15 @@ use std::fs::File;
 use std::io::Result;
 use std::sync::atomic::{AtomicBool, Ordering::Relaxed};
 
-pub(crate) const UTIME_NOW: i64 = 1_073_741_823;
-pub(crate) const UTIME_OMIT: i64 = 1_073_741_822;
+cfg_if::cfg_if! {
+    if #[cfg(target_pointer_width = "64")] {
+        pub(crate) const UTIME_NOW: i64 = 1_073_741_823;
+        pub(crate) const UTIME_OMIT: i64 = 1_073_741_822;
+    } else {
+        pub(crate) const UTIME_NOW: i32 = 1_073_741_823;
+        pub(crate) const UTIME_OMIT: i32 = 1_073_741_822;
+    }
+}
 
 /// Wrapper for `utimensat` syscall, however, with an added twist such that `utimensat` symbol
 /// is firstly resolved (i.e., we check whether it exists on the host), and only used if that is

--- a/crates/wasi-common/yanix/src/sys/linux/mod.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/mod.rs
@@ -8,8 +8,23 @@ use crate::dir::SeekLoc;
 use std::io::Result;
 
 impl SeekLoc {
+    #[cfg(target_pointer_width = "64")]
     pub unsafe fn from_raw(loc: i64) -> Result<Self> {
         let loc = loc.into();
+        Ok(Self(loc))
+    }
+    #[cfg(target_pointer_width = "32")]
+    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
+        // The cookie (or `loc`) is an opaque value, and applications aren't supposed to do
+        // arithmetic on them or pick their own values or have any awareness of the numeric
+        // range of the values. They're just supposed to pass back in the values that we
+        // give them. And any value we give them will be convertable back to `long`,
+        // because that's the type the OS gives them to us in. So return an `EINVAL`.
+        use std::convert::TryInto;
+        use std::io::Error;
+        let loc = loc
+            .try_into()
+            .map_err(|_| Error::from_raw_os_error(libc::EINVAL))?;
         Ok(Self(loc))
     }
 }

--- a/crates/wasi-common/yanix/src/sys/linux/mod.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/mod.rs
@@ -3,28 +3,3 @@ pub(crate) mod fadvise;
 pub(crate) mod file;
 pub(crate) mod filetime;
 pub(crate) mod utimesat;
-
-use crate::dir::SeekLoc;
-use std::io::Result;
-
-impl SeekLoc {
-    #[cfg(target_pointer_width = "64")]
-    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
-        let loc = loc.into();
-        Ok(Self(loc))
-    }
-    #[cfg(target_pointer_width = "32")]
-    pub unsafe fn from_raw(loc: i64) -> Result<Self> {
-        // The cookie (or `loc`) is an opaque value, and applications aren't supposed to do
-        // arithmetic on them or pick their own values or have any awareness of the numeric
-        // range of the values. They're just supposed to pass back in the values that we
-        // give them. And any value we give them will be convertable back to `long`,
-        // because that's the type the OS gives them to us in. So return an `EINVAL`.
-        use std::convert::TryInto;
-        use std::io::Error;
-        let loc = loc
-            .try_into()
-            .map_err(|_| Error::from_raw_os_error(libc::EINVAL))?;
-        Ok(Self(loc))
-    }
-}

--- a/crates/wasi-common/yanix/src/sys/linux/utimesat.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/utimesat.rs
@@ -1,4 +1,5 @@
 use crate::filetime::FileTime;
+use crate::filetime::FileTimeExt;
 use crate::from_success_code;
 use std::fs;
 use std::io::Result;
@@ -26,19 +27,16 @@ pub fn utimesat(
     let fd = unsafe { libc::openat(dirfd.as_raw_fd(), p.as_ptr(), flags) };
     let f = unsafe { fs::File::from_raw_fd(fd) };
     let (atime, mtime) = get_times(atime, mtime, || f.metadata().map_err(Into::into))?;
-    let times = [to_timeval(atime), to_timeval(mtime)];
+    let times = [to_timeval(atime)?, to_timeval(mtime)?];
     from_success_code(unsafe { libc::futimes(f.as_raw_fd(), times.as_ptr()) })
 }
 
-/// Converts `filetime::FileTime` to `libc::timeval`. This function was taken directly from
-/// [filetime] crate.
-///
-/// [filetime]: https://github.com/alexcrichton/filetime/blob/master/src/unix/utimes.rs#L93
-fn to_timeval(ft: filetime::FileTime) -> libc::timeval {
-    libc::timeval {
-        tv_sec: ft.seconds(),
-        tv_usec: (ft.nanoseconds() / 1000) as libc::suseconds_t,
-    }
+/// Converts `filetime::FileTime` to `libc::timeval`.
+fn to_timeval(ft: filetime::FileTime) -> Result<libc::timeval> {
+    Ok(libc::timeval {
+        tv_sec: ft.seconds_checked()?,
+        tv_usec: (ft.nanoseconds_checked()? / 1000) as libc::suseconds_t,
+    })
 }
 
 /// For a provided pair of access and modified `FileTime`s, converts the input to

--- a/crates/wasi-common/yanix/src/sys/linux/utimesat.rs
+++ b/crates/wasi-common/yanix/src/sys/linux/utimesat.rs
@@ -34,8 +34,8 @@ pub fn utimesat(
 /// Converts `filetime::FileTime` to `libc::timeval`.
 fn to_timeval(ft: filetime::FileTime) -> Result<libc::timeval> {
     Ok(libc::timeval {
-        tv_sec: ft.seconds_checked()?,
-        tv_usec: (ft.nanoseconds_checked()? / 1000) as libc::suseconds_t,
+        tv_sec: ft.seconds_()?,
+        tv_usec: (ft.nanoseconds_() / 1000) as libc::suseconds_t,
     })
 }
 

